torchvision/io/video.py

```diff
@@ -0,0 +1,146 @@
+import av
+import gc
+import torch
+import numpy as np
+import math
+
+
+# PyAV has some reference cycles
+_CALLED_TIMES = 0
+_GC_COLLECTION_INTERVAL = 20
+
+
+def write_video(filename, video_array, fps):
+    """
+    Writes a 4d tensor in [T, H, W, C] format in a video file
+
+    Arguments:
+        filename (str): path where the video will be saved
+        video_array (Tensor[T, H, W, C]): tensor containing the individual frames,
+            as a uint8 tensor in [T, H, W, C] format
+        fps (Number): frames per second
+    """
+    video_array = torch.as_tensor(video_array, dtype=torch.uint8).numpy()
+
+    container = av.open(filename, mode='w')
+
+    stream = container.add_stream('mpeg4', rate=fps)
+    stream.width = video_array.shape[2]
+    stream.height = video_array.shape[1]
+    stream.pix_fmt = 'yuv420p'
+
+    for img in video_array:
+        frame = av.VideoFrame.from_ndarray(img, format='rgb24')
+        for packet in stream.encode(frame):
+            container.mux(packet)
+
+    # Flush stream
+    for packet in stream.encode():
+        container.mux(packet)
+
+    # Close the file
+    container.close()
+
+
+def _read_from_stream(container, start_offset, end_offset, stream, stream_name):
+    global _CALLED_TIMES, _GC_COLLECTION_INTERVAL
+    _CALLED_TIMES += 1
+    if _CALLED_TIMES % _GC_COLLECTION_INTERVAL == _GC_COLLECTION_INTERVAL - 1:
+        gc.collect()
+
+    container.seek(start_offset, any_frame=False, backward=True, stream=stream)
+    frames = []
+    first_frame = None
+    for idx, frame in enumerate(container.decode(**stream_name)):
+        if frame.pts < start_offset:
+            first_frame = frame
+            continue
+        if first_frame and first_frame.pts < start_offset:
+            if frame.pts != start_offset:
+                frames.append(first_frame)
+            first_frame = None
+        frames.append(frame)
+        if frame.pts >= end_offset:
+            break
+    return frames
+
+
+def _align_audio_frames(aframes, audio_frames, ref_start, ref_end):
+    start, end = audio_frames[0].pts, audio_frames[-1].pts
+    total_aframes = aframes.shape[1]
+    step_per_aframe = (end - start + 1) / total_aframes
+    s_idx = 0
+    e_idx = total_aframes
+    if start < ref_start:
+        s_idx = int((ref_start - start) / step_per_aframe)
+    if end > ref_end:
+        e_idx = int((ref_end - end) / step_per_aframe)
+    return aframes[:, s_idx:e_idx]
+
+
+def read_video(filename, start_pts=0, end_pts=math.inf):
```

Comment:
Great question.

It is currently not visible to the user what the fps for a given video is.

I see two options:
- provide the fps somewhere, either by returning it in this function or by a separate helper function
- always enforce that the videos that are returned follow a particular fps. This might imply doing some expensive postprocessing / frame interpolation maybe?

At least the temporal video data would be consistent.

Thoughts?

Reply:
I feel like the two options are not (necessarily) mutually exclusive? 

1.  "provide the fps somewhere, either by returning it in this function or by a separate helper function"
This would be incredibly useful as that has been the main reproducibility issue of the formerly mentioned repo -- relatively non-standard encoding. Maybe we could (following torchaudio example) have `read_video()` return `(video, audio), stats` where stats would be a list/dict/tensor with video fps and audio sampling rate? It should be a part of the stream codec as [VideoCodecContext.framerate](https://docs.mikeboers.com/pyav/develop/api/video.html).

2. "always enforce that the videos that are returned follow a particular fps. This might imply doing some expensive postprocessing / frame interpolation maybe?"
This might be very slow could cause issues/artefacts within the video. We could leave that one out for future releases? Video frame sampling rate could be a relatively straightforward way to get a similar functionality that would allow some control without the need for expensive operations.